### ISSUE TITLE
Proposal: tie income/expense accounts directly to a cost center

### ISSUE THREAD

**Issue by jniles** [third-culture-software/bhima#5900]
This proposal extends and supersedes https://github.com/IMA-WorldHealth/bhima/issues/5858

I've been working with the Vanga database to try and create a database of cost centers tied to the income and expense accounts so that we can automatically infer costs when these accounts are used.   The configuration process is laborious for a simple objective.  All I need to know if given an account_id, what is the cost center associated with it?

Let's take the case of "Chirurgie".  In the current system, here is what I need to do:

1. Create the Service named "Chirurgie".
2. Create a principal cost center named "Chirurgie".  Assign the service to this cost center.
3. Create an account reference named "Chirurgie PC". Assign all "Chirurgie" income accounts to it (there is only one).
4. Create an account reference named "Chirurgie CC". Assign all "Chirurgie" cost accounts to it (there are two, maybe more depending on how you model it)
4. Return to the principal cost center "Chirurgie".  Assign the "Chirurgie CC" to the assigned cost references.  Assign "Chirurgie PC" to the assigned profit references.  I don't know what turnover means in this case, but I chose "autres".

Here is what I'd like to be able to do:

1. Create the service "Chirurgie" (and automatically create a cost center https://github.com/IMA-WorldHealth/bhima/issues/5854)
2. Go to the account page and add the "Chirurgie" center to all the "Chirurgie" income and expense accounts (usually, 1 - 5 accounts)

----

There are a couple of advantages to this system:

1. Firmly establishes that an account can have one and only one cost/revenue center.
2. Simplifies the logic to ensure we only attach cost/revenue centers to income/expense account types.
3. The associated cost/revenue center will be visible in the chart of accounts to easily audit.
4. Looking up the cost center associated with the account is very quick.

The disadvantages:

1. Incompatible with the previous way of doing things.
2. ?

**Comment by jmcameron** [comment]
I think the simpler we make it, the better.  So I think the new approach you suggest makes a lot of sense.   The fact that it is not compatible with the previous way of doing things is a secondary concern.   Although the work that was done in the past on fee centers was good, I think it is unnecessarily complicated.   I think that it is important that we do the hard work now to use a simpler approach.  As we try to do more and more financial planning things (eg, break-even analysis), we are much more likely to succeed in the time we have if if the system is less complicated.

My understanding of the proposed approach and our planning on step-down and break-even analysis makes me think that our proposed approach is flexible enough to do whatever we want to to.   For instance I think we could do the prior 'distribution key" approach using the proposed approachs by making another version of the step-down algorithm that just does a straight distributions.  Most of things we are developing should be able to handle it with minor modifications.   We can certainly do the proposed approaches in parallel with the old distribution key approach and then streamline things later once we have things working.

I thought of some questions/thoughts about your proposal:
- Will we always create cost centers before services?   We may need some flexibility (especially in legacy systems) to create services first (or deal with existing services).  This could be handled by supporting changing the cost centers when we edit services.
- I think we may need a simple way to associate cost centers with specific accounts.  Perhaps a hierarchical approach where top-level account families tag all their child accounts?
- We also need an easy way to identify which accounts do not have cost centers associated with them.
- We need a function somewhere that will tell us that all accounts have cost centers associated with them.  Whenever we do any type of cost center analysis, we will need to check to make sure that all accounts have cost centers associated with them before proceeding.

**Comment by jniles** [comment]
@jmcameron thanks for your comments.  Addressing them inline.

> Will we always create cost centers before services? We may need some flexibility (especially in legacy systems) to create services first (or deal with existing services). This could be handled by supporting changing the cost centers when we edit services.

To automatically allocate values on stock exit, payroll, and invoicing, we'll need each service to be associated with a cost center.  So, I think that legacy databases will need to be updated with cost centers for their services.  Alternatively, we implement an `enterprise_setting` that turns on and off the "lookup and associate costs" functionality.  That is doable as well.

> I think we may need a simple way to associate cost centers with specific accounts. Perhaps a hierarchical approach where top-level account families tag all their child accounts?

I don't mind implementing a way to have the parent assign cost centers to the children, recursively.  However, I think we should store one and only one id per account for two reasons:

1. We will be looking these up many many more times then we will be assigning them, so let's put it the same row as the account number, label, etc.   Then we can just pull out the cost center directly with a `SELECT` on the account id.
2. We need to ensure that an account cannot have two cost centers assigned to it.

In my mind, this is an advanced nice-to-have feature, not a strict requirement.

> We also need an easy way to identify which accounts do not have cost centers associated with them.
```sql
SELECT * FROM account WHERE account_type IN ('income', 'expense') AND cost_center_id IS NULL;
```

> We need a function somewhere that will tell us that all accounts have cost centers associated with them. Whenever we do any type of cost center analysis, we will need to check to make sure that all accounts have cost centers associated with them before proceeding.

```sql
SELECT * FROM account WHERE cost_center_id IS NOT NULL;
```

**Comment by jniles** [comment]
This was closed in https://github.com/IMA-WorldHealth/bhima/pull/5967
